File: java/src/sliding_window/LC_1876_SubstringsOfSizeThreeDistinctChars.java
package sliding_window;

// https://leetcode.com/problems/substrings-of-size-three-with-distinct-characters
import java.util.HashMap;
import java.util.HashSet;

public class LC_1876_SubstringsOfSizeThreeDistinctChars {

  public int countGoodSubstrings(String s) {
    int substringCount = 0, k = 3, left = 0;
    HashMap<Character, Integer> map = new HashMap<>();

    for (int j = 0; j < s.length(); j++) {
      map.put(s.charAt(j), map.getOrDefault(s.charAt(j), 0) + 1);

      if (j - left == k) {
        char ch = s.charAt(left);
        map.put(ch, map.get(ch) - 1);
        if (map.get(ch) == 0) {
          map.remove(ch);
        }
        left++;
      }

      if (map.size() == k) {
        substringCount++;
      }
    }
    return substringCount;
  }

  public int countGoodSubstrings_firstSol(String s) {
    int substringCount = 0, k = 3;
    HashMap<Character, Integer> map = new HashMap<>();
    for (int i = 0; i < k && i < s.length(); i++) {
      map.put(s.charAt(i), map.getOrDefault(s.charAt(i), 0) + 1);
    }
    if (map.size() == k) {
      substringCount++;
    }

    for (int j = k; j < s.length(); j++) {
      map.put(s.charAt(j - k), map.get(s.charAt(j - k)) - 1);
      if (map.get(s.charAt(j - k)) == 0) {
        map.remove(s.charAt(j - k));
      }
      map.put(s.charAt(j), map.getOrDefault(s.charAt(j), 0) + 1);
      if (map.size() == k) {
        substringCount++;
      }
    }
    return substringCount;
  }

  public int countGoodSubstrings_usingSet(String s) {
    int substringCount = 0, k = 3, left = 0;
    HashSet<Character> hashSet = new HashSet<>();
    for (int j = 0; j < s.length(); j++) {
      while (hashSet.contains(s.charAt(j))) {
        hashSet.remove(s.charAt(left++));
      }

      hashSet.add(s.charAt(j));

      if (hashSet.size() == k) {
        substringCount++;
        hashSet.remove(s.charAt(left++));
      }
    }
    return substringCount;
  }
}

File: python/sliding_window/LC_1876_SubstringsOfSizeThreeDistinctChars.py
# https://leetcode.com/problems/substrings-of-size-three-with-distinct-characters/
class Solution:
    def countGoodSubstrings(self, s: str) -> int:
        substring_count = left = 0
        k = 3
        map = {}
        for right in range(len(s)):
            map[s[right]] = map.get(s[right], 0) + 1
            if right - left == k:
                ch = s[left]
                map[ch] -= 1
                if map[ch] == 0:
                    del map[ch]
            if len(map) == k:
                substring_count += 1
        return substring_count

    def countGoodSubstrings_using_set(self, s: str) -> int:
        substring_count = left = 0
        k = 3
        window = set()
        for right in range(len(s)):
            while s[right] in window:
                window.remove(s[left])
                left += 1
            window.add(s[right])
            if len(window) == k:
                substring_count += 1
                window.remove(s[left])
                left += 1
        return substring_count
